# A study of controlling the false discovery rate

In this Jupyter notebook, we briefly study the analysis presented by Benjamini and Hochberg (1995) in their seminal paper on controlling the false discovery rate.

## Load and print the Neuhaus et al. p-values

Here we first import the function _neuhaus_ from the module _data_, which can be used to return the p-values reported by Neuhaus and co-authors (1995). We then print the p-values with four decimal places using the new string formatting syntax (http://pyformat.info) and finally store them for further processing.

In [12]:
from data import neuhaus
pvals = neuhaus()
print ['{:.4f}'.format(p) for p in pvals]
%store pvals

['0.0001', '0.0004', '0.0019', '0.0095', '0.0201', '0.0278', '0.0298', '0.0344', '0.0459', '0.3240', '0.4262', '0.5719', '0.6528', '0.7590', '1.0000']
Stored 'pvals' (ndarray)


## Perform the classic Bonferroni correction

Our dataset consists of p-values $p_{1}, \dots, p_{15}$ corresponding to a family of hypotheses $H_{1}, \dots, H_{15}$. Since there are $N=15$ hypotheses, the new critical level that takes multiple testing into account is $\alpha_{\text{Bonferroni}} = \alpha/N = 0.05/15 = 0.00334$. This operation is readily performed by using the function _bonferroni_ from the module _fwer_.

In [19]:
%store -r pvals
from fwer import bonferroni
significant_pvals = bonferroni(pvals)
print zip(['{:.4f}'.format(p) for p in pvals], significant_pvals)

[('0.0001', True), ('0.0004', True), ('0.0019', True), ('0.0095', False), ('0.0201', False), ('0.0278', False), ('0.0298', False), ('0.0344', False), ('0.0459', False), ('0.3240', False), ('0.4262', False), ('0.5719', False), ('0.6528', False), ('0.7590', False), ('1.0000', False)]


The first three p-values are significant after the correction since $0.0001 < 0.0004 < 0.0019 < \alpha_{\text{Bonferroni}} = 0.00334$.

## Perform the Benjamini & Hochberg FDR correction

In [3]:
%store -r pvals
from adaptive import lsu
significant_pvals = lsu(pvals, q=0.05)
print zip(['{:.4f}'.format(p) for p in pvals], significant_pvals)

[('0.0001', True), ('0.0004', True), ('0.0019', True), ('0.0095', True), ('0.0201', False), ('0.0278', False), ('0.0298', False), ('0.0344', False), ('0.0459', False), ('0.3240', False), ('0.4262', False), ('0.5719', False), ('0.6528', False), ('0.7590', False), ('1.0000', False)]


The first p-value is significant since $0.0001 < 1\cdot \frac{0.05}{15} = 0.00334$, the second is significant since $0.0004 < 2\cdot\frac{0.05}{15} = 0.00667$, the third is significant since $0.0019 < 3\cdot\frac{0.05}{15} = 0.01$, **and** the fourth is significant since $0.0095 < 4\cdot\frac{0.05}{15} = 0.01334$. The additional discovery is due to a smaller false-negative rate.